### PR TITLE
mpv: switch to lua@5.1

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Silently fails to build against Lua 5.3.

Fixes https://github.com/Homebrew/homebrew-core/issues/21388.

### CODE REVIEW

**Comment by DomT4** [comment]
Well, it did work locally. Quite what is happening here on CI I don't know:
```
==> python3 waf configure --prefix=/usr/local/Cellar/mpv/0.27.0_3 --enable-zsh-comp --enable-libmpv-shared --enable-html-build --confdir=/usr/local/etc/mpv --datadir=/usr/local/Cellar/mpv/0.27.0_3/share/mpv --mandir=/usr/local/Cellar/mpv/0.27.0_3/share/man --docdir=/usr/local/Cellar/mpv/0.27.0_3/share/doc/mpv --zshdir=/usr/local/Cellar/mpv/0.27.0_3/share/zsh/site-functions
Failed to execute: python3
```

**Comment by DomT4** [comment]
Well, I know _what_. But as for why, 🤷‍♂️.

**Comment by ilovezfs** [comment]
mpv is maintained. do they really not have lua 5.3 compat now? :(

**Comment by DomT4** [comment]
Apparently not. I haven't checked whether it works if you hack `5.3` onto the list of "valid" `lua` versions it checks for.

**Comment by DomT4** [comment]
https://github.com/mpv-player/mpv/commit/9b59c175e389e0a0561312139c0861eba37e4719. The tone seems to have softened since then but still no 5.3 support in `master`.

**Comment by ilovezfs** [comment]
Let's just require lua@5.1 so we don't need to deprecate the without option which is barely used.

**Comment by DomT4** [comment]
Done 👍.

**Comment by ilovezfs** [comment]
now we'll see if CI :pig: vomits again.

**Comment by DomT4** [comment]
I'm offline for a bit after this so if it borks & you work out why feel free to just push onto the branch. Fingers crossed it doesn't though 🙄.

**Comment by ilovezfs** [comment]
@DomT4 it's a bug due to https://github.com/Homebrew/brew/pull/3479 which is kicking in even for things not satisfied by the system. CC @MikeMcQuaid

**Comment by ilovezfs** [comment]
Reverting in https://github.com/Homebrew/brew/pull/3535.

**Comment by MikeMcQuaid** [comment]
Fix PR opened in https://github.com/Homebrew/brew/pull/3536.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by sCreami** [comment]
@ilovezfs merge this please

**Comment by SteveJobzniak** [comment]
I just spoke to mpv's developers. They say that Lua 5.1 is preferred. 5.2 is acceptable but doesn't always work. 5.3 will not work (due to https://github.com/mpv-player/mpv/commit/9b59c175e389e0a0561312139c0861eba37e4719).

So please merge this "lua@5.1" fix. It's critical since December 5th when Homebrew's lua formula changed from 5.2 to 5.3...

**Comment by CounterPillow** [comment]
>mpv is maintained. do they really not have lua 5.3 compat now? :(

https://github.com/mpv-player/mpv/wiki/FAQ#Why_does_mpv_not_support_Lua_53

There's no point in breaking every single user script simply because something has a bigger version number. Additionally, LuaJIT refuses to support 5.3 too, IIRC.

**Comment by ilovezfs** [comment]
Our 5.3 is built with 5.2 compatibility, which mpv really should be compatible with at this point. I'd like to see an open issue about this upstream since saying "5.3's 5.2 compatibility doesn't actually work" isn't a tenable position unless they're engaging with lua upstream about something specific.

**Comment by SteveJobzniak** [comment]
@ilovezfs Well, if you read the linked commit (I posted it above), it seems to mostly be about how lua 5.3 handles integers, which requires rewrites of mpv to shuttle those numbers back/forth in Lua's new integer format. More work than they have time for now.

They literally say that 5.1 is the best version, 5.2 is ok with some problems, and 5.3 is no-go. So let's just merge this to fix the mpv formula?

If anyone wants to desperately help mpv rewrite the code to Lua 5.3's new way, then feel free to join their team and help out. Until then, let's merge this.

**Comment by ilovezfs** [comment]
@SteveJobzniak I've already made my position clear. There needs to be an open mpv issue about supporting 5.3's 5.2 compatibility. I'm not asking for compatibility with 5.3 built _without_ 5.2 compat at this point.

**Comment by CounterPillow** [comment]
@ilovezfs so do I understand this correctly: you're blocking the merge of this because there is no open issue about 5.3-with-5.2-compat on mpv's bug tracker? Since you're so knowledgeable about this topic, why don't you open it with the needed information pertinent to such support? And why is that even a prerequisite for this change to be merged?

Distros breaking things for their users just for passive-aggressive political games are not exactly appreciated on mpv's side.

**Comment by ilovezfs** [comment]
@CounterPillow I suggest you read or re-read our [Code of Conduct](https://github.com/Homebrew/homebrew-core/blob/master/CODE_OF_CONDUCT.md) and adjust your future communications accordingly.

**Comment by CounterPillow** [comment]
@ilovezfs I believe any violation of the CoC you've linked is not on my part, but by you reading things into what I wrote. Additionally, I'd appreciate it if you'd actually address my points, as opposed to what you perceive to be my tone. Again, here are the questions I asked, repeated:

1. Did I understand this correctly that you are opposed to this change, on the basis I mentioned?
2. Why do you not take the information you have on 5.3-5.2 compat up to upstream, as opposed to having them guess?
3. If the basis in 1. is accurate, why is it a prerequisite for the merge of this change?

**Comment by SteveJobzniak** [comment]
Let's just look at the facts:

- mpv's repo has stated that 5.1 is the best version, 5.2 has some incompatibilities, 5.3 is actively blocked by their source code.
- The 5.3 version of lua has an incompatible way of handling numbers, which will not be supported (too much work).
- Lua versions are not meant to be direct upgrades. 5.1 != 5.2 != 5.3. They are not meant to be backwards-compatible or binary-compatible.
- wm4, on IRC, said that supporting both 5.1 and 5.2 was hard enough. Adding 5.3 is not going to happen for a long time.
- wm4, on IRC, said that he distrusts 5.3's "5.2 emulation" mode and thinks it's a bad idea and does not want to waste developer time on it, since Lua is capable of installing multiple versions simultaneously and all they're asking for is their version of Lua: 5.1 (or at worst, 5.2).
- In short: Mpv was built for Lua 5.1. Solvable by merging this pull request.

**Comment by MikeMcQuaid** [comment]
> Distros breaking things for their users just for passive-aggressive political games are not exactly appreciated on mpv's side.

@CounterPillow It is not a passive aggressive political game to think that upstream software should have an issue opened to at least consider supporting the second-newer minor version of software. Homebrew doesn't want to package Lua 5.1 forever.

> Since you're so knowledgeable about this topic, why don't you open it with the needed information pertinent to such support?

Because this is open source and it's not a reasonable expectation for a Homebrew maintainer to have to do all the legwork themselves.

---

And from a now-deleted comment:

> Additionally, I'd appreciate it if you'd actually address my points, as opposed to what you perceive to be my tone.

Tone matters more than content. If your tone is perceived to be rude by Homebrew's maintainers: you need to fix it or you will not be welcome in discussions on our issue tracker. I support @ilovezfs's position on that.

**Comment by ilovezfs** [comment]
thanks Dom

**Comment by ilovezfs** [comment]
Upstream issue here: https://github.com/mpv-player/mpv/issues/5205

**Comment by DomT4** [comment]
Thanks for the merge ILZ.

I'm not going to blow open the debate again here, but I will note that if Lua 5.1 _ever_ breaks on macOS for any reason whatsoever upstream seem to have made it clear they will not issue another release on that branch, essentially forcing Homebrew to then become maintainers of the Lua 5.1 branch or suddenly drop it completely & likely a bunch of formulae that rely on it.

It's not an enormous problem today but that doesn't mean that will never change, and consequently package managers/Homebrew have a legitimate interest in encouraging upstreams to adopt practices that reduce the risk of sudden, sharp breakage. `mpv` is a maintained project that can adjust both itself & its users to new situations over time, and for better or worse by this point, Lua 5.1 is not.

**Comment by SteveJobzniak** [comment]
@DomT4 Just remember that it's not just a matter of simply rewriting mpv to use Lua 5.3. Doing so requires rewrites of the built-in _.lua scripts_ (to adapt them to 5.3's new number format), which in turn might break them on all machines (the vast majority of mpv users) that have Lua 5.1 and Lua 5.2 instead of 5.3.

For now, the better path seems to be to move to LuaJIT if possible, which is a less insane project. LuaJIT actually sticks to the standard 5.1 language and maintains backwards compatibility, and they strongly dislike Lua 5.3's breaking changes and will not be including them in LuaJIT.

We know that mpv can built against LuaJIT on Linux. The question is if that'll work on macOS. It probably should, since the C library for both Lua 5.1 and LuaJIT is identical, so it shouldn't matter to mpv which one it gets linked to.

**Comment by MikeMcQuaid** [comment]
20v0ks

**Comment by SteveJobzniak** [comment]
@MikeMcQuaid Love it. A chuckle of reason in this madness.
